### ISSUE TITLE
DropBox tokens expire after 4 hours

### ISSUE THREAD

**Issue by YouveGotMeowxy** [sebastianfeldmann/phpbu#362]
@sebastianfeldmann 

Any chance of getting this fixed pretty soon? The key expires every 4 hours now, meaning adding to a daily cron job always fails, because the key has expired. Then in order to get it working again the user has to go back to their dbox app page, regen an 4hr key, and replace them all in their .json config. For me right now that's 9 locations in the configs. :)

btw, just wanted to add, this is one of the greatest apps ever, lol. Once it's all setup and running smooth, it's soo awesome.

_Originally posted by @YouveGotMeowxy in https://github.com/sebastianfeldmann/phpbu/issues/356#issuecomment-1890908435_

**Comment by YouveGotMeowxy** [comment]
@sebastianfeldmann ok, thank you for any work on it. :)

I don't really know if this will help you out any, but I was just looking around about how to get a permanent token and came across this; maybe it might help ya out as well.

https://www.codemzy.com/blog/dropbox-long-lived-access-refresh-token

**Comment by YouveGotMeowxy** [comment]
4 months since this issue; any forward-movement happening? :)

**Comment by jmcdougal** [comment]
I came across this post as I was having the same issue as I use Dropbox as well for my backups.  I resolved it by getting the application access code as mentioned here:  https://www.codemzy.com/blog/dropbox-long-lived-access-refresh-token.  Using CURL, I retrieved the long lived refresh token using the application access code with a grant type of "authorization_code".  I then created another CURL request to Dropbox asking for the short lived access token that I store in a MySQL database.  To do that, you use the long lived refresh token from the first CURL request with the grant type of "refresh_token".

An additional PHP script programmatically rebuilds the phpbu.xml file inserting the current refresh token into the line of <option name="token" value="' . $access_token .'"/>.  Using file_put_contents at the end of this script, the previous phpbu.xml file is replaced with a copy that has the newly updated token.

The backup runs twice daily using multiple cron jobs.  The first gets the updated access token.  The second updates the phpbu.xml script and the third runs the backup.  Each cron job is separated by five minutes.  This could be simplified to one cron job and script, but works fine as is and I have moved on to other projects.

Anyway, thought this might be helpful as you navigate this issue.  Overall, PHPBU is an excellent tool.  Great job Sebastian!

**Comment by sebastianfeldmann** [comment]
With version 6.0.25 the Dropbox integration was fixed.

You need to create an App in your Dropbox AppConsole.
Generate an Access Token and provide `token`, `appKey` and `appSecret` in the phpbu configuration file.

```xml
<?xml version="1.0" encoding="UTF-8"?>
<sync type="dropbox">
  <!-- mandatory -->
  <option name="token" value="mycrazylongapitokenthatigotfromdropbox" />

  <!-- mandatory -->
  <option name="appKey" value="myAppKey" />

  <!-- mandatory -->
  <option name="appSecret" value="myAppSecret" />

  <!-- mandatory -->
  <option name="path" value="/some/dir" />
</sync>
```

**Comment by sebastianfeldmann** [comment]
Yes, version 6 requires php 8

I think the dependencies are PHP 8 only but I can check and try to back port the feature to the 5.0 release version.
But I'm not sure that's possible.

**Comment by YouveGotMeowxy** [comment]
ok! I was able to change to PHP 8.3 and got it to run. I'm getting a sync error now, but gonna look into it!

**Comment by YouveGotMeowxy** [comment]
My fault, I got it and it's working great! You just made my whole day, lol. thank you so much! <3

**Comment by YouveGotMeowxy** [comment]
:(

I guess I spoke too soon. It worked after I had generated the new superlongtoken, but then overnight when my cronjob ran the same command-line I woke up to the same error that token expired.

```
            "class": "phpbu\\App\\Backup\\Sync\\Exception",
            "message": "{\"error_summary\": \"expired_access_token\/...\", \"error\": {\".tag\": \"expired_access_token\"}}",
            "file": "phar:\/\/\/home\/user\/public_html\/dbbackup\/phpbu\/phpbu-6.0.26.phar\/Backup\/Sync\/Dropbox.php",
            "line": 138
```

I don't know which specific info you may need from me to help out, just let me know.

**Comment by sebastianfeldmann** [comment]
I think I have found a way with refresh tokens.
I will test it again tomorrow, it it works I will release a new version and a guide how to obtain a refresh token for your app.

**Comment by sebastianfeldmann** [comment]
It is finally fixed!
At least it works on my machine ;)

The thing you have to be aware of is that you have to provide a `RefreshToken` not a `AccessToken`.
in order to generate on you habe to make a Dropbox API call.

You can  use the [dropbox-helper](https://github.com/sebastianfeldmann/phpbu-dropbox-helper) package I created to make the call and receive your refresh key.

With that key that you should never share with anybody you can create short-lived api `AccessTokens`.

**Comment by YouveGotMeowxy** [comment]
Thank you for keeping on top of this @sebastianfeldmann  :)

I'm probably being a doofus but when I run the new script, I just get this:

```
bin/dropbox
Please enter the AppKey: q6**********
Please enter the AppSecret: ex*********

https://dropbox.com/oauth2/authorize?client_id=q6**********&response_type=code&token_access_type=offline' - Log in with Dropbox

Please enter the code: 
```

I don't get any return code to enter?

I tried logging into both that particular app's settings page as well as the main account page, but see no "Refresh Token" or "Code" to enter into the new script?

also, a couple other questions:

once I do get my refresh token, do I just delete the original line (I'm using json btw):

    "token": "sl.CA-jsuperlongstring",

and replace it with (in json version):

    <option name="refreshToken" value="MY_REFRESH_TOKEN" />

and from then on does that same refresh token always work, or do I need to repeat the helper script every so often and manually edit the token in my configs?

Sorry if I'm asking imbecilic questions: :p

**Comment by sebastianfeldmann** [comment]
Not at all, you have to go to the URL and retrieve the `AppCode` and put that in the prompt on the cli to finally get the `RefreshToken`.

**Comment by sebastianfeldmann** [comment]
I added the URL step to the `README` of the helper package.

**Comment by sebastianfeldmann** [comment]
Good idea, I will update the message 👍

**Comment by YouveGotMeowxy** [comment]
So far so good! WEWT!!! TYSM

I don't have many heroes, but you've just been added to my list! :p

I have them set to run nightly on my cron, if it poops out I'll letcha know! :)


Also, none of the following is very important, but would make for very nice feedback enhancements. It seems the backup script waits for all backups to complete before giving any user feedback whatsoever. In the case of backing up many things, including encrypting, sending to DB, etc., it can take a pretty long time, and visually the script/terminal just seems to "hang" with a blinking cursor. It could easily be confused with a freeze.

Maybe initially adding something like "performing backup", "encrypting", "sending to DB", etc. as it's performing each step, and then showing the little chart after each backup is complete, rather than a long pause with a big "dump" at the end? Feedback is always nice. :)

Just an idea! :)

**Comment by YouveGotMeowxy** [comment]
Last question tonight so I stop annoying you! lol

Has this been fixed yet? https://github.com/sebastianfeldmann/phpbu/issues/364

**Comment by YouveGotMeowxy** [comment]
poke @sebastianfeldmann :)

**Comment by YouveGotMeowxy** [comment]
poke @sebastianfeldmann ;)

**Comment by sebastianfeldmann** [comment]
Check comment @ #364
